Document new features in run retries, make difference between OSS and cloud clearer (#20105)
dagster-io/dagster: docs/content/deployment/run-retries.mdx

# Run Retries

> If you configure run retries, a new run will be kicked off whenever a run fails for any reason. Compared to [Op retries](/concepts/ops-jobs-graphs/op-retries), the maximum retry limit for run retries applies to the whole run instead of each individual Op. Run retries also handle the case where the run process crashes or is unexpectedly terminated.

Reviewer: ```suggestion
If you configure run retries, a new run will be kicked off whenever a run fails for any reason. Compared to [Op retries](/concepts/ops-jobs-graphs/op-retries), the maximum retry limit for run retries applies to the whole run instead of each individual op. Run retries also handle the case where the run process crashes or is unexpectedly terminated.
```

Sorry this was bugging me

## Configuration

How to configure a default maximum number of retries for all runs in your Dagster deployment depends on whether you're using Dagster Cloud or Dagster Open Source:

- **Dagster Cloud**: Use the [Dagster Cloud UI or the dagster-cloud CLI][cloud-deployment-settings]

> - **Dagster Open Source**: Use your instance's `dagster.yaml`

Author: @erinkcochran87 this is driving me nuts, do you have any idea why it doesn't like this? I copy-pasted it verbatim from the concurrency docs, but it appears incorrectly and mdx-format auto-corrects it to:
`
**Dagster Cloud**: Use the \[Dagster Cloud UI or the dagster-cloud CLI]\[cloud-deployment-settings]

> - **Dagster Cloud**: Use the [Dagster Cloud UI or the dagster-cloud CLI](/dagster-cloud/managing-deployments/managing-deployments#configuring-deployment-settings) to set a default maximum number of retries. Run retries do not need to be explicitly enabled.

Author: i was able to make mdx-format happy by using the URL instead of the 
`
[cloud-deployment-settings]: /dagster-cloud/managing-deployments/managing-deployments#configuring-deployment-settings
`
thing that the concurrency page uses

Reviewer: So, so weird. I can't wait til' we're rid of this. Glad you got it working tho!

> For example, the following will set a default maximum number of retries of 3 for all runs:

Reviewer: ```suggestion
For example, the following will set a default maximum number of retries of `3` for all runs:
```

Nit, sorry

By default, retries will re-execute from failure (tag value `FROM_FAILURE`). This means that any successful Ops will be skipped, but their output will be used for downstream Ops. If the `dagster/retry_strategy` tag is set to `ALL_STEPS`, all the Ops will run again.

> **Note:** `FROM_FAILURE` requires an IOManager that can access outputs from other runs. For example, on Kubernetes the [s3\_pickle_io_manager](/\_apidocs/libraries/dagster-aws#dagster_aws.s3.s3\_pickle_io_manager) would work but the [`FilesytemIOManager`](https://docs.dagster.io/\_apidocs/io-managers#dagster.FilesytemIOManager) would not, since the new run is in a new Kubernetes Job with a separate filesystem.

Reviewer: ```suggestion
**Note:** `FROM_FAILURE` requires an I/O manager that can access outputs from other runs. For example, on Kubernetes the <PyObject object="s3_pickle_io_manager" module="dagster_aws.s3.s3" /> would work but the <PyObject object="FilesytemIOManager" module="dagster" /> would not, since the new run is in a new Kubernetes Job with a separate filesystem.
```

I know this wasn't you, just fixing it while we're in here.

> ### Combining op retries and run retries

Reviewer: Thoughts on `Combining op and run retries`?

Author: done!

> By default, if a run fails due to an op failure and both op retries and run retries are enabled, the overlapping retries might cause the op to be retried more times than desired, since the op retry count will reset for each retried run.

Reviewer: This sentence is a lot to take in - thoughts on breaking it up?

```suggestion
By default, if a run fails due to an op failure and both op and run retries are enabled, the overlapping retries might cause the op to be retried more times than desired. This is because the op retry count will reset for each retried run.
```

> To prevent this from happening, you can configure run retries to only retry when the failure is due to a reason other than an op failure (for example, a crash or an unexpected termination of the run worker). This is controlled by the `run_retries.retry_on_asset_or_op_failure` setting, which defaults to true but can be overridden to false.

Reviewer: ```suggestion
To prevent this from happening, you can configure run retries to only retry when the failure is due to a reason other than an op failure. For example, a crash or an unexpected termination of the run worker. This is controlled by the `run_retries.retry_on_asset_or_op_failure` setting, which defaults to `true` but can be overridden to `false`.
```

Author: Tweaked a bit, lmk what you think

For example, the following configures run retries so that they ignore runs that failed due to a step failure:

```yaml
run_retries:
  enabled: true # Omit this key if using Dagster Cloud, since run retries are enabled by default
  max_retries: 3
  retry_on_asset_or_op_failure: false
```

> **Note:** Setting `retry_on_asset_or_op_failure` for false will only change retry behavior for runs on dagster version 1.6.7 or greater.

Reviewer: ```suggestion
**Note:** Setting `retry_on_asset_or_op_failure` to `false` will only change retry behavior for runs on Dagster version 1.6.7 or greater.
```
